dask_planner/src/sql/optimizer/filter_columns_post_join.rs

```diff
@@ -0,0 +1,678 @@
+                        fetch: l.fetch,
+                        input: Arc::new(previous_step.clone()),
+                    });
+                }
+                LogicalPlan::Distinct(_) => {
+                    return_plan = LogicalPlan::Distinct(Distinct {
+                        input: Arc::new(previous_step.clone()),
+                    });
+                }
+                LogicalPlan::Aggregate(a) => {
+                    return_plan = LogicalPlan::Aggregate(Aggregate {
+                        input: Arc::new(previous_step.clone()),
+                        group_expr: a.group_expr.clone(),
+                        aggr_expr: a.aggr_expr.clone(),
+                        schema: a.schema.clone(),
+                    });
+                }
+                LogicalPlan::Sort(s) => {
+                    return_plan = LogicalPlan::Sort(Sort {
+                        expr: s.expr.clone(),
+                        input: Arc::new(previous_step.clone()),
+                        fetch: s.fetch,
+                    });
+                }
+                LogicalPlan::Subquery(_) => {
+                    return_plan = LogicalPlan::Subquery(Subquery {
+                        subquery: Arc::new(previous_step.clone()),
+                    });
+                }
+                _ => {
+                    warn!("Skipping optimizer rule 'FilterColumnsPostJoin'");
+                    return Ok(plan.clone());
+                }
+            }
+            previous_step = &return_plan;
+        }
+
+        Ok(return_plan)
+    } else {
+        Ok(plan.clone())
+    }
+}
+
+fn filter_post_join_columns(
+    post_join_columns: &HashSet<Expr>,
+    schemas: &Vec<&Arc<DFSchema>>,
+) -> HashSet<Expr> {
+    let mut result = HashSet::new();
+
+    let mut valid_qualifiers: Vec<&str> = vec![];
+    for dfschema in schemas {
+        let dfschema_fields = dfschema.fields();
+        for field in dfschema_fields {
+            let qualifier = field.qualifier();
+            if let Some(q) = qualifier {
+                valid_qualifiers.push(q);
+            }
+        }
+    }
+
+    for column in post_join_columns {
+        if let Expr::Column(c) = column {
+            let column_qualifier = &c.relation;
+            if let Some(q) = column_qualifier {
+                if valid_qualifiers.contains(&q.as_str()) {
+                    result.insert(column.clone());
+                }
+            } else {
+                result.insert(column.clone());
+            }
+        }
+    }
+    result
+}
+
+fn insert_post_join_columns(post_join_columns: &mut HashSet<Expr>, inserted_columns: &Vec<Expr>) {
+    for column in inserted_columns {
+        if let Some(exprs) = get_column_name(column) {
+            for expr in exprs {
+                post_join_columns.insert(expr);
+            }
+        }
+    }
+}
+
+fn get_column_name(column: &Expr) -> Option<Vec<Expr>> {
+    // TODO: Can we use to_columns() or expr_to_columns() here?
```

Comment:
Right now we use DataFusion 14.0.0, but once we update to 15.0.0 we might be able to simplify this function by just using https://github.com/apache/arrow-datafusion/blob/4538912db3c913220bfa0b251b6ab62842886145/datafusion/expr/src/expr.rs#L701